string method
string topic
---
string[] protocols
string[] endpoints
